module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,67 @@
+#include "compression.hpp"
+
+#include <iostream>
+#include <string>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    std::array<std::array<uint8_t, height>, width>& uncompressed) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressed;
+    compressed.reserve(height * width);
+
+    for (const auto& row : uncompressed) {
+        auto rowIterator = row.cbegin();
+        uint8_t value = *rowIterator;
+        uint8_t currentValueCounter = 1;
+        rowIterator++;
+
+        while (rowIterator != row.cend()) {
+            if (*rowIterator == value) {
+                currentValueCounter++;
+            } else {
+                compressed.push_back(std::make_pair(value, currentValueCounter));
+                value = *rowIterator;
+                currentValueCounter = 1;
+            }
+            rowIterator++;
+        }
+        // and for last value in a row:
+        compressed.push_back(std::make_pair(value, currentValueCounter));
+    }
```

Comment:
Try to reconfigure your loop, to call push_back only once.

Reply:
@nauka-programowania-MA I think the code is clear enough.
Currently he inserts on two events:
a) change of element
b) end of row
If he was to change that he would either have to :
a) insert once and update value in vector on every occurence
b) do an edge case for the last element in row in both if {} and else {} part of his loop.

Reply:
I tried to refactor this loop make "push_back" occur only once, but that simplification entails complication in conditionary instruction, as @piezol said. It comes to the same thing.